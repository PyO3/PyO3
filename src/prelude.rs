//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
