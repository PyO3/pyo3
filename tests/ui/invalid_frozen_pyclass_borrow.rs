use pyo3::prelude::*;

#[pyclass(frozen)]
pub struct Foo {
    #[pyo3(get)]
    field: u32,
}

#[pymethods]
impl Foo {
    fn mut_method(&mut self) {}
}

fn borrow_mut_fails(foo: Py<Foo>, py: Python) {
    let borrow = foo.bind(py).borrow_mut();
}

#[pyclass(subclass)]
struct MutableBase;

#[pyclass(frozen, extends = MutableBase)]
struct ImmutableChild;

fn borrow_mut_of_child_fails(child: Py<ImmutableChild>, py: Python) {
    let borrow = child.bind(py).borrow_mut();
}

fn py_get_of_mutable_class_fails(class: Py<MutableBase>) {
    class.get();
}

fn pyclass_get_of_mutable_class_fails(class: &Bound<'_, MutableBase>) {
    class.get();
}

#[pyclass(frozen)]
pub struct SetOnFrozenClass {
    #[pyo3(set)]
    field: u32,
}

fn main() {}
