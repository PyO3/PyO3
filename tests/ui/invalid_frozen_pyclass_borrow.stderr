error: cannot use `#[pyo3(set)]` on a `frozen` class
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:39:12
   |
39 |     #[pyo3(set)]
   |            ^^^

error[E0271]: type mismatch resolving `<Foo as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:12:19
   |
12 |     fn mut_method(&mut self) {}
   |                   ^^^^^^^^^ expected `False`, found `True`
   |
note: required by a bound in `extract_pyclass_ref_mut`
  --> src/impl_/extract_argument.rs
   |
   | pub fn extract_pyclass_ref_mut<'a, 'py: 'a, T: PyClass<Frozen = False>>(
   |                                                        ^^^^^^^^^^^^^^ required by this bound in `extract_pyclass_ref_mut`

error[E0271]: type mismatch resolving `<Foo as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:10:1
   |
10 | #[pymethods]
   | ^^^^^^^^^^^^ expected `False`, found `True`
   |
note: required by a bound in `PyRefMut`
  --> src/pycell.rs
   |
   | pub struct PyRefMut<'p, T: PyClass<Frozen = False>> {
   |                                    ^^^^^^^^^^^^^^ required by this bound in `PyRefMut`
   = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<Foo as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:16:31
   |
16 |     let borrow = foo.bind(py).borrow_mut();
   |                               ^^^^^^^^^^ expected `False`, found `True`
   |
note: required by a bound in `pyo3::Bound::<'py, T>::borrow_mut`
  --> src/instance.rs
   |
   |     pub fn borrow_mut(&self) -> PyRefMut<'py, T>
   |            ---------- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = False>,
   |                    ^^^^^^^^^^^^^^ required by this bound in `Bound::<'py, T>::borrow_mut`

error[E0271]: type mismatch resolving `<ImmutableChild as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:26:33
   |
26 |     let borrow = child.bind(py).borrow_mut();
   |                                 ^^^^^^^^^^ expected `False`, found `True`
   |
note: required by a bound in `pyo3::Bound::<'py, T>::borrow_mut`
  --> src/instance.rs
   |
   |     pub fn borrow_mut(&self) -> PyRefMut<'py, T>
   |            ---------- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = False>,
   |                    ^^^^^^^^^^^^^^ required by this bound in `Bound::<'py, T>::borrow_mut`

error[E0271]: type mismatch resolving `<MutableBase as PyClass>::Frozen == True`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:30:11
   |
30 |     class.get();
   |           ^^^ expected `True`, found `False`
   |
note: required by a bound in `pyo3::Py::<T>::get`
  --> src/instance.rs
   |
   |     pub fn get(&self) -> &T
   |            --- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = True> + Sync,
   |                    ^^^^^^^^^^^^^ required by this bound in `Py::<T>::get`

error[E0271]: type mismatch resolving `<MutableBase as PyClass>::Frozen == True`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:34:11
   |
34 |     class.get();
   |           ^^^ expected `True`, found `False`
   |
note: required by a bound in `pyo3::Bound::<'py, T>::get`
  --> src/instance.rs
   |
   |     pub fn get(&self) -> &T
   |            --- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = True> + Sync,
   |                    ^^^^^^^^^^^^^ required by this bound in `Bound::<'py, T>::get`
